module #4. Authentication & Spring Security/oauth2-resource-server/src/main/java/com/epam/esm/config/SecurityConfig.java

```diff
@@ -0,0 +1,120 @@
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.HttpStatus;
+import org.springframework.security.access.PermissionEvaluator;
+import org.springframework.security.access.expression.method.DefaultMethodSecurityExpressionHandler;
+import org.springframework.security.access.expression.method.MethodSecurityExpressionHandler;
+import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.http.SessionCreationPolicy;
+import org.springframework.security.oauth2.jwt.JwtDecoder;
+import org.springframework.security.oauth2.jwt.JwtDecoders;
+import org.springframework.security.web.AuthenticationEntryPoint;
+import org.springframework.security.web.SecurityFilterChain;
+import org.springframework.security.web.access.AccessDeniedHandler;
+import org.springframework.util.MimeTypeUtils;
+
+import java.io.IOException;
+
+@Configuration
+@EnableWebSecurity
+@EnableMethodSecurity
+public class SecurityConfig {
+
+    private static final String[] TAG_ENDPOINTS = {"/api/tags", "/api/tags/**"};
+    private static final String[] USER_ENDPOINTS = {"/api/users", "/api/users/**"};
+    private static final String[] CERTIFICATE_ENDPOINTS = {"/api/certificates", "/api/certificates/**"};
+    private static final String[] ORDER_ENDPOINTS = {"/api/orders", "/api/orders/**"};
+    private static final String ADMIN = "ADMIN";
+    private static final String USER_ORDERS_ENDPOINT = "/api/orders/users/*";
+
+    private final String issuer;
+    private final JwtAuthenticationTokenConverter jwtAuthenticationTokenConverter;
+    private final PermissionEvaluator permissionEvaluator;
+    private final ExceptionMessageI18n i18n;
+
+    public SecurityConfig(OAuth2ResourceServerProperties properties,
+        JwtAuthenticationTokenConverter jwtAuthenticationTokenConverter, PermissionEvaluator permissionEvaluator,
+        ExceptionMessageI18n i18n) {
+        this.issuer = properties.getJwt().getIssuerUri();
+        this.jwtAuthenticationTokenConverter = jwtAuthenticationTokenConverter;
+        this.permissionEvaluator = permissionEvaluator;
+        this.i18n = i18n;
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        http
+
+            .sessionManagement(session -> session
+                .sessionCreationPolicy(SessionCreationPolicy.STATELESS))
+
+            .authorizeHttpRequests(request -> request
+                // Guest
+                .requestMatchers(HttpMethod.GET, CERTIFICATE_ENDPOINTS).permitAll()
+
+                // User
+                .requestMatchers(HttpMethod.GET, TAG_ENDPOINTS).authenticated()
```

Comment:
I`d suggest to use @PreAuthorize instead of request matches, the more endpoints you have -> the bigger this config would be, and it would be easier to make an error